Acceptance testing (#1822)
quay/claircore: docs/howto/acceptance_testing.md

|--------|---------|---------------|
| `affected` | Product is vulnerable | MUST appear in analyzer results as affected |
| `not_affected` | Product is known to not be affected | MUST appear in analyzer results as not-affected |
| `absent` | Product should not appear (fixed/filtered) | MUST NOT appear in analyzer results |

Lines starting with `#` are treated as comments.

### Uploading Fixtures

Use the `main.go` tool in the acceptance directory to create fixtures:

```bash
# Navigate to the acceptance directory
cd acceptance

# Create a fixture
go run -mod=mod main.go create \
    -image registry.io/namespace/image@sha256:abc123... \
    -tag my-fixture-name \
    -vex path/to/vex.json \
    -manifest path/to/expected.csv \
    -repo your.registry.io/namespace/your-repo

# You can attach multiple VEX documents
go run -mod=mod main.go create \
    -image registry.io/namespace/image@sha256:abc123... \
    -tag multi-vex-fixture \
    -vex vex1.json \
    -vex vex2.json \
    -vex vex3.json \
    -manifest expected.csv \
    -repo your.registry.io/your-repo

# Verify the fixture was created
go run -mod=mod main.go list -image your.registry.io/your-repo:my-fixture-name

> ```

Reviewer: These should use command's path, now that it's split out.

#### Create Options

| Flag | Default | Description |
|------|---------|-------------|
| `-image` | (required) | Source image reference with digest |
| `-tag` | (required) | Tag name for the fixture |
| `-vex` | (required) | Path to VEX document (repeatable) |
| `-manifest` | (required) | Path to expected results CSV |
| `-repo` | `quay.io/projectquay/clair-fixtures` | Target repository |
| `-platform` | `linux/amd64` | Platform to copy (use `all` for multi-arch) |

## Running Tests

### Running Integration Tests

The integration tests automatically download and run embedded PostgreSQL
binaries - no database setup required:

```bash
cd acceptance
go test -tags=integration -v ./...
```

### Using Local OCI Layout

For testing without a registry, fixtures can be loaded from local OCI Layout
directories:

```go
acceptance.Run(ctx, t, auditor, []string{
    "ocidir:///path/to/local/fixture",
})
```

## Testing with Claircore

Most teams will use the built-in claircore analyzer to test their images. Here's
a complete, copy-paste-able test file:

```go

> //go:build integration

Reviewer: ```suggestion
```

}
```

Example implementation:

```go
type MyAuditor struct {}

func (a *MyAuditor) Audit(ctx context.Context, t testing.TB, ref string, csafDocs [][]byte) ([]acceptance.Result, error) {
    // 1. Index the image at 'ref' to get packages/SBOMs
    packages, err := a.indexImage(ctx, ref)
    if err != nil {
        return nil, err
    }
    
    // 2. Parse the CSAF documents
    for _, doc := range csafDocs {
        vulns, err := a.parseCSAF(doc)
        if err != nil {
            return nil, err
        }
        // 3. Match packages against vulnerabilities
    }
    
    // 4. Return results in the expected format, the auditor will do the comparison.
    return results, nil
}

func TestMyAuditor(t *testing.T) {
    ctx := context.Background()
    auditor := &MyAuditor{/* ... */}
    
    acceptance.Run(ctx, t, auditor, []string{
        "quay.io/projectquay/clair-fixtures:python-test",
    })
}
```

## Media Types

| Type | Media Type | Description |
|------|------------|-------------|
| VEX | `application/csaf+json` | CSAF/VEX vulnerability document |
| Expected Results | `application/vnd.com.redhat.container.acceptancetest.v1+csv` | Test expectations CSV |

> Compressed variants (`.../+zstd`) are also supported.

Reviewer: ```suggestion
Compressed variants (`...+zstd`) are also supported.
```

I think inserting a slash is needlessly complicated.

|-------|-------------|
| `document.tracking.id` | The CVE/tracking ID (must match CSV) |
| `product_tree.branches[].product.product_id` | Product identifier (must match CSV) |
| `product_tree.branches[].product.product_identification_helper.purl` | Package URL for matching |
| `vulnerabilities[].product_status.known_affected` | List of affected product IDs |
| `vulnerabilities[].product_status.known_not_affected` | List of not-affected product IDs |

### Expected Results CSV

The expected results file is a CSV with three columns (no header row):

```csv
tracking_id,product_id,status
```

Example:

```csv
CVE-2024-1234,mypackage-1.2.3,affected
CVE-2024-5678,otherpackage-2.0.0,not_affected
CVE-2024-9999,fixedpackage-3.0.0,absent
```

| Status | Meaning | Test Behavior |
|--------|---------|---------------|
| `affected` | Product is vulnerable | MUST appear in analyzer results as affected |
| `not_affected` | Product is known to not be affected | MUST appear in analyzer results as not-affected |
| `absent` | Product should not appear (fixed/filtered) | MUST NOT appear in analyzer results |

Lines starting with `#` are treated as comments.

### Uploading Fixtures

> Use the `main.go` tool in the acceptance directory to create fixtures:

Reviewer: ```suggestion
Use the `test/acceptance/cmd/fixture` tool to create fixtures:
```

```bash
# Navigate to the acceptance directory
cd acceptance

> 

Reviewer: ```suggestion
```

    -vex path/to/vex.json \
    -manifest path/to/expected.csv \
    -repo your.registry.io/namespace/your-repo

# You can attach multiple VEX documents
go run ./test/acceptance/cmd/fixture create \
    -image registry.io/namespace/image@sha256:abc123... \
    -tag multi-vex-fixture \
    -vex vex1.json \
    -vex vex2.json \
    -vex vex3.json \
    -manifest expected.csv \
    -repo your.registry.io/your-repo

# Verify the fixture was created
go run ./test/acceptance/cmd/fixture list -image your.registry.io/your-repo:my-fixture-name
```

#### Create Options

| Flag | Default | Description |
|------|---------|-------------|
| `-image` | (required) | Source image reference with digest |
| `-tag` | (required) | Tag name for the fixture |
| `-vex` | (required) | Path to VEX document (repeatable) |
| `-manifest` | (required) | Path to expected results CSV |
| `-repo` | `quay.io/projectquay/clair-fixtures` | Target repository |
| `-platform` | `linux/amd64` | Platform to copy (use `all` for multi-arch) |

## Running Tests

### Running Integration Tests

The integration tests automatically download and run embedded PostgreSQL
binaries - no database setup required:

```bash

> cd acceptance

Reviewer: ```suggestion
cd test/acceptance
```

    matcherPool := testpostgres.TestMatcherDB(ctx, t)

    // Create the claircore auditor
    auditor, err := acceptance.NewClaircoreAuditor(ctx, t, &acceptance.ClaircoreConfig{
        IndexerPool: indexerPool,
        MatcherPool: matcherPool,
        Platform:    "linux/amd64",
    }, &http.Client{Timeout: 2 * time.Minute})
    if err != nil {
        t.Fatalf("create auditor: %v", err)
    }
    t.Cleanup(func() { auditor.Close(ctx) })

    // Run tests against your fixtures
    acceptance.Run(ctx, t, auditor, []string{
        "quay.io/your-team/your-fixtures:app-v1",
        "quay.io/your-team/your-fixtures:app-v2",
    })
}
```

To run:

```bash
go test -tags=integration -v -run TestMyImages ./...
```

The test will:
1. Pull each fixture from the registry
2. Index the image with claircore
3. Load the VEX documents attached to the image
4. Match vulnerabilities against the indexed packages
5. Compare results to the expected results CSV
6. Report any mismatches or missing vulnerabilities

## Implementing a Custom Auditor

If you're building your own vulnerability analyzer (not using claircore), implement
the `Auditor` interface:

> ```go

Reviewer: Might be better to replace this code block with a `pkg.go.dev` link, so that it doesn't rot.

Author: Good idea

type Auditor interface {

>     Audit(ctx context.Context, t testing.TB, ref string, csafDocs [][]byte) ([]Result, error)

Reviewer: ```suggestion
    Audit(ctx context.Context, t testing.TB, ref string, csafDocs iter.Seq[io.Reader]) ([]Result, error)
```

}

type Result struct {
    TrackingID string                      // CVE ID from the VEX document
    ProductID  string                      // Product ID from the VEX product tree
    Status     fixtures.VulnerabilityStatus // affected, not_affected, absent.
    Package    string                      // All matched package
}
```

Example implementation:

```go
type MyAuditor struct {}

func (a *MyAuditor) Audit(ctx context.Context, t testing.TB, ref string, csafDocs [][]byte) ([]acceptance.Result, error) {
    // 1. Index the image at 'ref' to get packages/SBOMs
    packages, err := a.indexImage(ctx, ref)
    if err != nil {
        return nil, err
    }
    
    // 2. Parse the CSAF documents
    for _, doc := range csafDocs {
        vulns, err := a.parseCSAF(doc)
        if err != nil {
            return nil, err
        }
        // 3. Match packages against vulnerabilities
    }
    
    // 4. Return results in the expected format, the auditor will do the comparison.
    return results, nil

> }

Reviewer: This could be an actual go file and included, so that the interface implementation could get typechecked.
